### DISCUSSION TITLE
Comfy Controller - The extension to build custom GUIs on each workflow

### DISCUSSION THREAD

**Discussion by LukeG89** [Comfy-Org/ComfyUI#5408] [Show and tell]
# Comfy Controller
### [Comfy Controller](https://github.com/chrisgoringe/cg-controller) (cg-controller) is developed by @chrisgoringe, the same developer of [cg-use-everywhere](https://github.com/chrisgoringe/cg-use-everywhere)

![1 4_preview](https://github.com/user-attachments/assets/433cdf12-7499-43d1-be1a-207413d830a9)

Join [Discord](https://discord.gg/p3vDHn3n)

Check the [Discussions](https://github.com/chrisgoringe/cg-controller/discussions)

See the [Milestones](https://github.com/chrisgoringe/cg-controller/milestones)

Open a [Bug report](https://github.com/chrisgoringe/cg-controller/issues/new) or a [Feature request](https://github.com/chrisgoringe/cg-controller/issues/new)

# Installation: Search "Comfy Controller" 🔎

![comfy_controller_install](https://github.com/user-attachments/assets/d27f0545-7673-4cd2-9b84-600c68a92ebf)

**Comment by JorgeR81** [comment]
**Proof of concept !**
---

This is NOT a mockup. 

It's just a screenshot taken while using the `dev` branch 

This version allows multiple controller instances. 
With multiple controllers, the user could assemble a whole custom GUI, on top of the graph canvas, for each workflow.

And we can hide this GUI, in with a toggle button in the Top Bar ( next to the Manager ), or via keybind: `c`

But for you to build a whole GUI, it would only be practical to do so, once we have panel "snapping" ( https://github.com/chrisgoringe/cg-controller/issues/228 ), witch may happen in version `1.3` ( https://github.com/chrisgoringe/cg-controller/milestone/26 )



---

This is the default workflow.

![p1b](https://github.com/user-attachments/assets/cc906517-1374-4483-9a07-d36c36115972)

---

Here the `Image Preview` controller is collapsed, so that we can see the workflow.

You can see the collapsed controller panel, in the top left corner of the graph canvas area.
It's also possible to interact with the workflow in this configuration.

![p2b](https://github.com/user-attachments/assets/051ff213-e987-4d0f-9949-32e99bf77cb8)

---

One downside of completely covering the graph canvas is not seeing the workflow progress.
But we can enable the rgthree's progress bar.

![rg3](https://github.com/user-attachments/assets/33759dbc-1e06-44c2-b669-fcf68750b6c3)

**Comment by rollingcookies** [comment]
Guys, look at how it's already implemented in Invoke AI, I've been using it for a while and the Workflows tab is how the UX in ComfyUI should look like and be organized. And it's drop dead convenient!

The panel on the left is all the assigned fields in separately selected nodes (not the whole node, just the individual fields) and also in the panel for a particular field you can specify a different name. In this way a Dashboard of specific workflow controls is assembled.

The right side of the screen is occupied by the workflow or a big picture of the generation process. This state is easily toggled with a button. The generation process on the fly shows what is happening in the sampler, so that you can stop the process in the early stages of generation. Or this area can also be used to preview images from a gallery of already generated images.

Thus, we get a simple and convenient uniform interface, where everything is always in its place. And whatever workflow we open, it doesn't take long to figure out, because there is a “Dashboard” mode with the control panel on the left and the output image / series of output images on the right in an area that takes up most of the screen.

**Reply by JorgeR81** [reply]
You are welcome to make suggestions in the extension's repo ( https://github.com/chrisgoringe/cg-controller ), to talk directly to the developer.
You can open a discussion or if you have a more specific idea you can open an issue, for a feature request. 

I can try to answer some things:

---

> The panel on the left is all the assigned fields in separately selected nodes (not the whole node, just the individual fields)

In the future it may be possible to select the widgets you want in the Controller, instead of adding the whole node. 
( https://github.com/chrisgoringe/cg-controller/issues/143 )

Right now, this can be done, by connecting a Primitive node. 
And adding the Primitive to the Controller instead.

---

> and also in the panel for a particular field you can specify a different name

The Controller just uses the same `group`, `node` and `widget` names you have in the nodes graph.
We thought it would be less confusing, this way.

----

> The right side of the screen is occupied by the workflow or a big picture of the generation process. 

You will be able to add multiple windows, and place them where you want in the canvas.
And you can display any nodes you want in these windows.

So you should be able to get similar layouts to the ones used by other programs.  
You can try this in the `dev` version.

---

> The generation process on the fly shows what is happening in the sampler, so that you can stop the process in the early stages of generation. 

If you add a Sampler node to the Controller, it will also show the in-progress preview, in the Controller.  

In Comfy UI, you can also create a group node with: `KSampler` + `Vae Decode` + `Image Preview`.
This way, the final image will show up in the same node as the in progress preview. 
You can use the `SamplerCustomAdvanced` node, so that there are no other widgets.

And you can add this group node to the Controller.
Then you can place it in the "big picture area", on the right side.

----

> Thus, we get a simple and convenient uniform interface, where everything is always in its place. And whatever workflow we open, it doesn't take long to figure out, because there is a “Dashboard” mode with the control panel on the left and the output image / series of output images on the right in an area that takes up most of the screen.

Yeah, I think that's the way most of the workflows will be organized.  
You can create a layout like that in a few minutes, save it in a workflow and always use it.

I think most shared workflows will also have that layout.
If not, you should be able to convert them easily, just by moving a few windows.

Once we have a better idea of how people will use this extension, we could have other features, related to specific layouts.

**Reply by rollingcookies** [reply]
> > and also in the panel for a particular field you can specify a different name
> 
> The Controller just uses the same `group`, `node` and `widget` names you have in the nodes graph. We thought it would be less confusing, this way.

Ok, that was just my opinion, which coincided with the InvokeAI developers' opinion, the panel on the left is the top level of the workflow interaction interface that the user forms on the fly. So if the panel on the left says “Control Structure” and this references the “End Percent” widget in Control Net in the workflow, it will be more clear to the end user in the context of use what they are dealing with than just “End Percent”. The user will not need to get into the noodles, he will immediately understand what he is dealing with having a simple “Dashboard” view of the workflow. If now it is realized only through the Primitive node, I hope that in time it will be possible to simply click on the name of the field in the right pane and edit the name. To avoid user confusion, for edited names in nodes in the workflow, it is possible to mark the edited name in some way: write in small font the changed name under the widget name, and when hovering over the widget in the left pane highlight this field in the node in the workflow, and vice versa - when hovering over the widget in the workflow highlight this field in the left pane. Or somehow else let the user know that this field is present in the left pane. 

Regarding the possibility to change the filling of the right area with any panels, you are right, it will be even more flexible for different users and tasks to be able to create different interface configurations with different panels.

And I also highly recommend trying out how Workflows are implemented in InvokeAI!

**Reply by JorgeR81** [reply]
> So if the panel on the left says “Control Structure” and this references the “End Percent” widget in Control Net in the workflow, it will be more clear to the end user in the context of use what they are dealing with than just “End Percent”. 

In the Controller, the widgets are always associated with a node ( even if you only show a single widget per node ).
And you can change the node's name, to use as context / reference.

![ct01](https://github.com/user-attachments/assets/e808e3ef-58af-434b-be25-7536a4fdb117)


---

>  I hope that in time it will be possible to simply click on the name of the field in the right pane and edit the name. 

Do you mean changing the widget's name ?

You can do that in ComfyUI, if you make a group node. 
And this way you can have all the ControlNet widget's, under the same "header", with the name you want.

We want the Controller to mirror what's in the workflow.

![ct02](https://github.com/user-attachments/assets/d5f09ee8-8f68-4fcd-8b23-51ba8b35e3f6)

**Reply by rollingcookies** [reply]
What I have in mind is already implemented in Invoke! :)

1) When you hover over a Node in the Workflow or a widget in the Controller, they are equally highlighted in both areas:
![2024-11-02_13-17-09](https://github.com/user-attachments/assets/87a15bf0-d662-4ed9-8d00-69e4a6963e30)

2) You can easily add any node field as a widget in the Controller:
![2024-11-02_13-28-04](https://github.com/user-attachments/assets/40dedeaf-38c0-44ad-8bb7-df4fca14bd3d)

3) You can easily change the name of a field in a node or on a Controller widget!
![2024-11-02_13-17-28](https://github.com/user-attachments/assets/f9327d39-1c13-481b-9919-b9a5198974aa)
![2024-11-02_13-17-59](https://github.com/user-attachments/assets/cb4921cf-8fef-4a56-a0aa-100568623d95)

4) You can easily remove a widget from a controller from any area: Workflow or Controller
![2024-11-02_13-18-45](https://github.com/user-attachments/assets/45c4414e-274d-48f7-b795-9e2c03d28bee)
![2024-11-02_13-19-31](https://github.com/user-attachments/assets/f9ae8cfc-ca9b-4580-a920-2d9ec0be3e76)


5) You can see detailed information about each widget in the Controller
![2024-11-02_13-19-19](https://github.com/user-attachments/assets/e603dc82-1585-40b2-996d-7a7a7b722a73)

**Reply by JorgeR81** [reply]
> You can easily change the name of a field in a node or on a Controller widget!

But when you change the name in one place, it's updated in the other, right ?
To have the same name in both places.

ComfyUI does not allow you to change the widget name ( unless you have a group node ).
The Controller can't change that.

ComfyUI will add a feature that allows to hide widgets in the node.
- https://github.com/Comfy-Org/litegraph.js/pull/250

The Controller will probably mirror this feature.

**Reply by LukeG89** [reply]
@rollingcookies I've never used Invoke, seems a pretty nice software! 
I think that some workarounds are needed to reach similar "Invoke-like" functionalities in ComfyUI (they are different software after all) and Controller can't overcome core limitations.

Some features can definitely be requested in the [frontend repo](https://github.com/Comfy-Org/ComfyUI_frontend) (e.g. "rename widgets") and then be implemented in Controller.

@chrisgoringe What do you think about it?

**Reply by JorgeR81** [reply]
@rollingcookies, I think I understand the argument for wanting to add individual widgets ?

You want to sort widgets, independently from nodes ( e.g. having all Loaders together: Checkpoint, ControlNet,  IPAdapter, ... ).

Is that right ?

----

![invoke](https://github.com/user-attachments/assets/f04f8085-bb54-4643-8c72-7be4f1dcd117)


---

In ComfyUI, there are individual nodes to load models.
So you can also add them individually and place them together in the Controller.
Sometimes, I also place them together in the node's graph.

![cui](https://github.com/user-attachments/assets/907fa8bb-5a1f-45c4-9105-c3fbd9f7c230)

**Reply by rollingcookies** [reply]
> > You can easily change the name of a field in a node or on a Controller widget!
> 
> But when you change the name in one place, it's updated in the other, right ? To have the same name in both places.

Exactly, but that would be ideal, I realize there are limitations at the moment. So different options could be considered to get closer to what Invoke has.

> ComfyUI does not allow you to change the widget name ( unless you have a group node ). The Controller can't change that.

What can we do with the appearance of nodes in the workflow using the Controller? Is it possible to add some controls or buttons, add text above or below a widget, highlight a separate widget with a frame or background highlighting, catch cursor hovering on the whole node and a separate widget?

> ComfyUI will add a feature that allows to hide widgets in the node.
> 
> * [Hidden & advanced widgets Comfy-Org/litegraph.js#250](https://github.com/Comfy-Org/litegraph.js/pull/250)
> 
> The Controller will probably mirror this feature.

I don't think that's quite what's needed

**Reply by rollingcookies** [reply]
> @rollingcookies, I think I understand the argument for wanting to add individual widgets ?
> 
> You want to sort widgets, independently from nodes ( e.g. having all Loaders together: Checkpoint, ControlNet, IPAdapter, ... ).
> 
> Is that right ?
> 
> ![invoke](https://private-user-images.githubusercontent.com/72734038/382474500-f04f8085-bb54-4643-8c72-7be4f1dcd117.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1NTg2MjEsIm5iZiI6MTczMDU1ODMyMSwicGF0aCI6Ii83MjczNDAzOC8zODI0NzQ1MDAtZjA0ZjgwODUtYmI1NC00NjQzLThjNzItN2JlNGYxZGNkMTE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDE0Mzg0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1ZjJkNjlkOGMzMWY4YzkwN2UzZjdjMzI4Njc0ZjAxYTVhYzhiN2JmNjg3NTQxMTFlODJhM2ZjMDY4NmE2MDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.twfkvVYCuPPCa9u7BbpbaR7bdNAkFRZMRJmKtiV8vmw)

That's right! See those handles on the left side of the widgets in the Controller, where you can move the widgets inside the controller and change their order? It's very convenient to have a separate block in the Controller that is just the widget you need from any workflow node. You don't need to add all the widgets of a node in the Controller, it overloads the interface and is not necessary for the end user of the workflow, but only what the user will use in the work. The same goes for the name of the widget itself. The user needs a clear name to work with, not just a technical term taken out of context.


> In ComfyUI, there are individual nodes to load models. So you can also add them individually and place them together in the Controller. Sometimes, I also place them together in the node's graph.
> 
> ![cui](https://private-user-images.githubusercontent.com/72734038/382475729-907fa8bb-5a1f-45c4-9105-c3fbd9f7c230.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1NTg2MjEsIm5iZiI6MTczMDU1ODMyMSwicGF0aCI6Ii83MjczNDAzOC8zODI0NzU3MjktOTA3ZmE4YmItNWExZi00NWM0LTkxMDUtYzNmYmQ5ZjdjMjMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDE0Mzg0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkYjQzZmMyYmQ5NmNkMzAwMTY4M2MxMDY2YTcwM2RkZWQ1OGIyYzExOGI4NzMyMzQwYTAxYzYzMDFjOWQ5MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qFgq-UvZjyHxSO8zN6ychDhmzHdsB98_nF2tNQBFh_Q)

Organizing the right order in the Controller is an important and useful task, it is often used to simplify access to information.

I also wanted to write, as you mentioned about node groups and primitive nodes for adding necessary widgets to the Controller. 

In general, I try not to use node groups in workflows, because it makes it difficult to understand such workflows, if you don't use them for a while, and then you need to change something, it can be difficult to understand such a workflow. So the best practice is to organize a workflow without nested nodes as simple and easy to read as possible so that you can quickly understand it after a while. 

I use primitives as needed, but you must agree that simply adding a widget from a node to the Controller and easily changing its name is not the same as first converting a widget to a port, then creating a separate primitive, renaming it and then adding the primitive to the Controller. There is a difference, right? )

Ideally, I think the way Invoke is done now is just super convenient. It is clear that there are limitations, we just need to understand how to get closer to this ideal solution using the features that are available now or that can be realized by adding functionality to the main ComfyUI.

**Reply by JorgeR81** [reply]
I don't know if it's possible to add individual widgets to the Controller.

But if it were possible, would this work for you ?

- All widgets are independent, you can move or remove them as you want.
- You need 2 rows per widget, like on InvokeAI
- The top row has the node's name. It's the same name for all widgets that came from that node.
- The bottom row has the slider and the widget's name. 
- You can only change the node's name, in the nodes graph. 
- The widget's name can only be changed if/when ComfyUI allows to do this. 

It could look like this:

![p02](https://github.com/user-attachments/assets/91e86ffc-76aa-4350-818d-14e58ee4fcbe)

**Reply by rollingcookies** [reply]
> I don't know if it's possible to add individual widgets to the Controller.

@chrisgoringe Is it possible, for example, in the context menu of adding a node to the controller, to make it possible to add only individual widgets? 
![menu](https://github.com/user-attachments/assets/355bae0b-c9b9-4bdf-ab03-77e336350e3c)


> But if it were possible, would this work for you ?

Partially yes )

> * All widgets are independent, you can move or remove them as you want.
> * You need 2 rows per widget, like on InvokeAI
> * The top row has the node's name. It's the same name for all widgets that came from that node.
> * The bottom row has the slider and the widget's name.
> * You can only change the node's name, in the nodes graph.
> * The widget's name can only be changed if/when ComfyUI allows to do this.
> 
> It could look like this:
> 
> ![p02](https://private-user-images.githubusercontent.com/72734038/382495482-91e86ffc-76aa-4350-818d-14e58ee4fcbe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1NzMzOTUsIm5iZiI6MTczMDU3MzA5NSwicGF0aCI6Ii83MjczNDAzOC8zODI0OTU0ODItOTFlODZmZmMtNzZhYS00MzUwLTgxOGQtMTRlNThlZTRmY2JlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDE4NDQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzNjdlZTUxOWQ3ODMwZDYxOTU2MWM2NmZhNDVlNWEzNTFhMDEyNTczMjJmMWM0M2ExNDAzMjRhZTQ3ZDQ5ZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YqbI6kj81xPPLT6RMXLPqODDVPVKlwK_yOeuqkg1g9M)

Is it possible not to show the Node name and the real name of the widget, but hide them in the icon ( i ), as it is done in Invoke?
When a widget is added to the Controller, the default data in the header is the widget name from the nodes graph, but if you click on the header, you can specify any data for the widget name in the Controller? Only the slider with the value will remain in the second line.

Also, in order to easily identify which widget belongs to which node on the graph, we need to solve the second problem, in several ways (it is better to implement several at once), which of them are possible:
- when hovering over a widget in the Controller, display a frame around it and also in the node graph display a frame around the widget on the node
- make another icon ( > ) on the widget in the Controller, which will perform quick focus (zooming and positioning of the node graph on the desired node with simultaneous selection of it (as if we clicked on it with the mouse).
- on the node in the node graph on the right or left side display a small marker (the same size as a port, but, for example, in the form of a blue arrow ( > )), which will mean that this widget is added to the controller. It would also be desirable to be able to highlight this widget with a frame when hovering the mouse cursor over this marker in the Controller
- When hovering over an icon ( i ) on a widget in the Controller, display information in a pop-up tooltip: “Start percent: 0.01 in Depth Controlnet, id: 102” 
- Hovering over the icon ( > ) next to the widget on the graph node will display a tooltip containing information: "Control Structure added as a widget", where “Control Structure” is the name of the widget in the Controller
- There may be some more ideas... I will add more later.

**Reply by chrisgoringe** [reply]
I'm not sure I see the utility of being able to move individual widgets around - conceptually, all the widgets from a given node "belong" together.

But being able to hide some widgets on a node in the controller (probably treating them as advanced), yes

**Reply by chrisgoringe** [reply]
Displaying the node in the workflow when you hover over it is on the roadmap

**Reply by LukeG89** [reply]
> I'm not sure I see the utility of being able to move individual widgets around - conceptually, all the widgets from a given node "belong" together.

@rollingcookies The "primitive" workaround is a good option if you only need a couple of widgets, and converting them to input has become easy thanks to the "fast input connection" (implemented by Pythongosssss a month ago)


https://github.com/user-attachments/assets/3ee4591c-d602-4ddf-904c-a89f20d0c173

**Reply by rollingcookies** [reply]
> > I'm not sure I see the utility of being able to move individual widgets around - conceptually, all the widgets from a given node "belong" together.
> 
> @rollingcookies The "primitive" workaround is a good option if you only need a couple of widgets, and converting them to input has become easy thanks to the "fast input connection" (implemented by Pythongosssss a month ago)
> 
>  Primitive_Workaround.mp4

I wanted to try the new versions, which are still in development, but I can not install either Dev branch or 1.3 - after installation just hangs in the panel button “Toggle Controller”, which can not be clicked, you do not know what is the reason?

Thanks for the new “primitive”, installed, it will certainly be more convenient! But must have features, still - is the ability to simply edit the name of the widget without a separate edit the name of the node (+ for some reason double click on the title of the primitive does not work and need to go through the context menu of the node > Title to edit mode)

**Reply by chrisgoringe** [reply]
Haven't seen that hang - could you press f12 and see if there are any errors being reported in the JavaScript console?

**Reply by LukeG89** [reply]
> I wanted to try the new versions, which are still in development, but I can not install either Dev branch or 1.3 - after installation just hangs in the panel button “Toggle Controller”, which can not be clicked, you do not know what is the reason?

You installed it using git? If so, open terminal inside cg-controller folder and run this command: `git branch --show-current`
If the current branch is 1.3, run this command to update it: `git pull`

> But must have features, still - is the ability to simply edit the name of the widget without a separate edit the name of the node

I don't know if this can be implemented, since you can't do it in ComfyUI.
I think this feature needs to be requested in [ComfyUI-frontend](https://github.com/Comfy-Org/ComfyUI_frontend) (or maybe [litegraph.js](https://github.com/Comfy-Org/litegraph.js)), you can do it if you want.

> (+ for some reason double click on the title of the primitive does not work and need to go through the context menu of the node > Title to edit mode)

Today I've literally opened a feature request about that! 😆
- https://github.com/Comfy-Org/ComfyUI_frontend/issues/1413

**Reply by JorgeR81** [reply]
> for some reason double click on the title of the primitive does not work and need to go through the context menu of the node > Title to edit mode)

When you only need to input a number, you can also use FLOAT or INT nodes, like these ones form the `Impact Pack`.

![ipk](https://github.com/user-attachments/assets/e0b49c3a-ec73-4ab6-b3bd-d11f74e0e036)

**Reply by rollingcookies** [reply]
> Haven't seen that hang - could you press f12 and see if there are any errors being reported in the JavaScript console?

Updated ComfyUI and all nodes, now when I install Dev (git clone -b dev https://github.com/chrisgoringe/cg-controller) or Main (git clone https://github.com/chrisgoringe/cg-controller) from the repository, I just get buttons in the panel that are clicked, but nothing happens, even in the main version

![2024-11-04_15-58-23](https://github.com/user-attachments/assets/9af493ed-d3ed-4a59-b5cf-f02c34cee702)

Fortunately, I saved yesterday's versions of Main and Dev and yesterday's Main displays the controller panel with no problems:
 
![2024-11-04_16-06-07](https://github.com/user-attachments/assets/b736ff30-910c-43a1-9745-40ad64950793)

**Reply by LukeG89** [reply]
@rollingcookies I think you just need to create a New Controller with the right-click menu. You can create multiple windows.

![new_controller_option](https://github.com/user-attachments/assets/96d6cdfa-f092-404b-b7bc-8c6438b9169a)

**Reply by rollingcookies** [reply]
> > I wanted to try the new versions, which are still in development, but I can not install either Dev branch or 1.3 - after installation just hangs in the panel button “Toggle Controller”, which can not be clicked, you do not know what is the reason?
> 
> You installed it using git? If so, open terminal inside cg-controller folder and run this command: `git branch --show-current` If the current branch is 1.3, run this command to update it: `git pull`

I tried downloading the zip file and unzipping to both the "cg-controller" folder and the "cg-controller-1.3" folder but nothing had any effect, then tried installing with the "git clone -b dev command https://github.com/chrisgoringe/cg-controller" - is this the correct command to install a non-main branch?

**Reply by JorgeR81** [reply]
> is this the correct command to install a non-main branch?

This is the easiest way to try the dev version:

- Open the "cg-controller" folder, in the "custom_nodes" folder
- Open the cmd line
- To update, run: `git pull`
- Then run: `git switch dev`
- If, later on, you need to go back, run: `git switch main`

---

You don't need to use `git clone`

**Reply by LukeG89** [reply]
@rollingcookies Let's start over! Delete all the cg-controller folders you have and follow this instructions:

- Go to your custom_nodes folder
- Open terminal
- Run this command: `git clone -b dev https://github.com/chrisgoringe/cg-controller.git`

You should have the latest beta Controller installed.
If you don't see any Controller window, this could be the reason: https://github.com/comfyanonymous/ComfyUI/discussions/5408#discussioncomment-11142581

**Reply by JorgeR81** [reply]
@LukeG89, in the images we can see that @rollingcookies, already has the Manager.

If the Controller was installed via Manager, maybe the easier way, is to use `git switch`, once we have a "cg-controller" folder.
There is no need to delete folders or use `git clone`

**Reply by JorgeR81** [reply]
@rollingcookies, now we only have `main` and `dev` branches. The others were removed.

- In the `main` branch, we can add multiple floating windows ( it was the previous `1.2` branch )

- In the `dev` branch, we have Group Tabs - Bypass Group - Live generation preview in Preview Image ( previous `1.3` branch  )


----

To check your current branch, run: `git branch --show-current`

Like this:

![cmd](https://github.com/user-attachments/assets/2946b02a-76ee-4136-bf6a-24875a88aee5)

**Reply by rollingcookies** [reply]
Everything works guys, thanks!

**Reply by LukeG89** [reply]
@rollingcookies Yay! 🎉 
New feedbacks are welcome!
If you encounter any bugs or you have a suggestion, please [open a new issue here](https://github.com/chrisgoringe/cg-controller/issues/new)!

**Reply by rollingcookies** [reply]
@LukeG89  Of course they will 😄, I've long wanted a solution inside ComfyUI that avoids resorting to the glitchy and ugly SwarmUI 😄!

**Comment by JorgeR81** [comment]
Some development updates:
----

Here is a 3 stage workflow, with FaceDetailer and ControlNet.

All nodes have the image on top of the widgets, so that all images are the same level. 

![z2](https://github.com/user-attachments/assets/92b177f7-6d75-4646-8561-bb56c2103e2e)


---

Here I added a `Load Image` node, to another Controller, to show a larger preview of the final image.

![z1](https://github.com/user-attachments/assets/65810389-39c7-4d07-9848-569444d3a598)

---

The generation preview from of all the samplers in the previous stages, will be shown in this large Image area.

The in-progress preview in the `Preview Image` node, is working only in branch `1.3`, so is still experimental.

![v1_3d](https://github.com/user-attachments/assets/9046ef97-1bb7-436c-b276-388b4aa31255)


---

If I colapse the Controller for the large image area, you can see the whole workflow.

That small blue group at the right, is used to feed the large image area.
It has an `Image Preview` node, with a darker grey color ( the same color shown in the large image area ).

![z3](https://github.com/user-attachments/assets/bd2cf35d-9df4-4368-a245-ea830cb186f0)

**Comment by JorgeR81** [comment]
Group Nesting was just added to ComfyUI ( https://github.com/Comfy-Org/ComfyUI_frontend/pull/1416 )

Now we can have groups within groups !

And it seems to work very well with the Controller.

![gn](https://github.com/user-attachments/assets/7375a1f6-fb8c-415d-bdf2-8c8206d88760)
